src/utils/recipeUtils.ts

```diff
@@ -57,24 +61,38 @@ export const getRequiredServiceIds = (recipe: RecipeDefinition): RegistryId[] =>
       .filter((serviceId) => serviceId !== PIXIEBRIX_SERVICE_ID)
   );
 
-export const getContainedExtensionPointTypes = (
-  recipe: RecipeDefinition
-): ExtensionPointType[] => {
-  const extensionPointTypes = new Set<ExtensionPointType>();
-
-  for (const definition of Object.values(recipe.definitions)) {
-    if (definition.kind !== "extensionPoint") {
-      continue;
-    }
-
-    const extensionPointDefinition =
-      definition.definition as ExtensionPointDefinition;
-    const extensionPointType = extensionPointDefinition?.type;
+const getExtensionPointType = async (
+  extensionPoint: ExtensionDefinition,
+  recipe?: RecipeDefinition
+): Promise<ExtensionPointType | null> => {
+  // Look up the extension point in recipe inner definitions first
+  if (recipe.definitions) {
+    const definition: ExtensionPointDefinition = recipe.definitions[
+      extensionPoint.id
+    ].definition as ExtensionPointDefinition;
+    const extensionPointType = definition?.type;
 
     if (extensionPointType) {
-      extensionPointTypes.add(extensionPointType);
+      return extensionPointType;
     }
   }
 
-  return [...extensionPointTypes];
+  // If no inner definitions, look up the extension point in the registry
+  const extensionPointFromRegistry = await extensionPointRegistry.lookup(
+    extensionPoint.id as RegistryId
+  );
+
+  return (extensionPointFromRegistry?.kind as ExtensionPointType) ?? null;
+};
+
+export const getContainedExtensionPointTypes = async (
+  recipe: RecipeDefinition
+): Promise<ExtensionPointType[]> => {
+  const extensionPointTypes = await Promise.all(
+    recipe.extensionPoints.map(async (extensionPoint) =>
+      getExtensionPointType(extensionPoint, recipe)
+    )
+  );
+
+  return uniq(extensionPointTypes.filter((type) => type !== null));
```

Comment:
Can also use lodash `compact()` here to remove nulls if you want:
```suggestion
  return uniq(compact(extensionPointTypes));
```

Reply:
Sweet!